PowerEditor/src/MISC/Common/Common.cpp

```diff
@@ -1331,23 +1331,37 @@ int nbDigitsFromNbLines(size_t nbLines)
 
 generic_string getDateTimeStrFrom(const generic_string& dateTimeFormat, const SYSTEMTIME& st)
 {
-	generic_string dateTimeStr = dateTimeFormat;
-	dateTimeStr = stringReplace(dateTimeStr, TEXT("Y"), std::to_wstring(st.wYear));
-	wchar_t buf[3];
-	_snwprintf(buf, sizeof(buf), TEXT("%02d"), st.wMonth);
-	dateTimeStr = stringReplace(dateTimeStr, TEXT("M"), buf);
-
-	_snwprintf(buf, sizeof(buf), TEXT("%02d"), st.wDay);
-	dateTimeStr = stringReplace(dateTimeStr, TEXT("D"), buf);
-
-	_snwprintf(buf, sizeof(buf), TEXT("%02d"), st.wHour);
-	dateTimeStr = stringReplace(dateTimeStr, TEXT("h"), buf);
-
-	_snwprintf(buf, sizeof(buf), TEXT("%02d"), st.wMinute);
-	dateTimeStr = stringReplace(dateTimeStr, TEXT("m"), buf);
+	const TCHAR* localeName = LOCALE_NAME_USER_DEFAULT;
+	const DWORD flags = 0;
+
+	constexpr int bufferSize = MAX_PATH;
+	TCHAR buffer[bufferSize] = {};
+	int ret = 0;
+
+	// 1. Remove format specifiers that we don't support:
+	//    - 'tt' that means AM/PM or 't' that means A/M
+	//    - 'g', 'gg' that means B.C. or A.D.
+	// Note: If 'tt' is supported it will need to be properly escaped
+	// to avoid conflict with 'M' date format that stands for month.
+	generic_string newFormat = dateTimeFormat;
+	auto it = std::remove_if(newFormat.begin(), newFormat.end(),
+		[](generic_string::value_type x) { return x == 't' || x == 'g'; });
+	newFormat.erase(it, newFormat.end());
+
+	// 2. Format the time (h/m/s/t/H).
+	ret = GetTimeFormatEx(localeName, flags, &st, newFormat.c_str(), buffer, bufferSize);
+	if (ret != 0)
+	{
+		// 3. Format the date (d/y/g/M).
+		// Now use the buffer as a format string to process the format specifiers not recognized by GetTimeFormatEx().
+		ret = GetDateFormatEx(localeName, flags, &st, buffer, buffer, bufferSize, nullptr);
+	}
 
-	_snwprintf(buf, sizeof(buf), TEXT("%02d"), st.wSecond);
-	dateTimeStr = stringReplace(dateTimeStr, TEXT("s"), buf);
+	if (ret != 0)
+	{
+		// 'ret' holds the actual length of the string
+		return generic_string(buffer, buffer + ret);
+	}
 
-	return dateTimeStr;
+	return {};
 }

```

Comment:
Dosn't the following simpler code work? If it does for any case, no need to remove `tt` & `gg`. I just don't like the extra processing.
```
generic_string getDateTimeStrFrom(const generic_string& dateTimeFormat, const SYSTEMTIME& st)
{
	const TCHAR* localeName = LOCALE_NAME_USER_DEFAULT;
	const DWORD flags = 0;

	constexpr int bufferSize = MAX_PATH;
	TCHAR buffer[bufferSize] = {};

	GetDateFormatEx(localeName, flags, &st, dateTimeFormat.c_str(), buffer, bufferSize, nullptr);
	GetTimeFormatEx(localeName, flags, &st, buffer, buffer, bufferSize);

	return buffer;
}
```

Reply:
Hmm... I see why you process time before date - due to `ddd`

Reply:
Right.
Processing time before date almost works.
Except for `tt`. If it inserts AM then `M` is considered as month specifier for date format.
So the string like `hh tt yy` will be formatted as `11 A9 21` instead of `11 AM 21` for `2021.09.02 11:00`.

Reply:
I can return `tt` processing like before but leave it empty for case when it's not defined in the current locale (as on your system).

Reply:
`tt` worked in my system if I use only GetTimeFormatEx without GetDateFormatEx.
Even with GetDateFormatEx called after, it still works - just M is replaced by a number (I didn't modify my settings).
So I guess there were some bug in your old commit.

Reply:
You can check this one:
https://ci.appveyor.com/project/donho/notepad-plus-plus/builds/40594850
It should work for `tt`.

Reply:
No, it doesn't.

![image](https://user-images.githubusercontent.com/90293/131884676-1f39e5d4-c126-4226-898d-95b473c08ea6.png)

Reply:
Even it worked, algorithm would be too complex to go into master.